src/ppx/knob.cpp

```diff
@@ -0,0 +1,300 @@
+        for (int i = 0; i < mChoices.size(); ++i) {
+            bool isSelected = (i == mIndex);
+            if (ImGui::Selectable(mChoices.at(i).c_str(), isSelected)) {
+                if (i != mIndex) { // A new choice is selected
+                    mIndex = i;
+                    if (mCallback)
+                        mCallback(mIndex);
+                }
+            }
+            if (isSelected) {
+                ImGui::SetItemDefaultFocus();
+            }
+        }
+        ImGui::EndCombo();
+    }
+}
+
+std::string KnobStrDropdown::GetFlagHelpText() const
+{
+    std::string choiceStr = "";
+    for (auto choice : mChoices) {
+        choiceStr += '\"' + choice + '\"' + "|";
+    }
+    if (!choiceStr.empty())
+        choiceStr.erase(choiceStr.size() - 1);
+
+    std::string flagHelpText = "--" + mFlagName + " <" + choiceStr + ">";
+    if (!mFlagDesc.empty()) {
+        flagHelpText += " : " + mFlagDesc;
+    }
+    return flagHelpText + "\n";
+}
+
+void KnobStrDropdown::UpdateFromFlag(const CliOptions& opts)
+{
+    std::string newVal = opts.GetExtraOptionValueOrDefault(mFlagName, GetStr());
+    SetDefault(newVal);
+}
+
+void KnobStrDropdown::SetIndex(size_t newI)
+{
+    if (!IsValidIndex(newI)) {
+        PPX_LOG_ERROR(mFlagName << " does not have this index in allowed choices: " << newI);
+        return;
+    }
+
+    // update mIndex and trigger mCallback
+    if (newI != mIndex) {
+        mIndex = newI;
+        if (mCallback)
+            mCallback(mIndex);
+    }
+}
+
+void KnobStrDropdown::SetIndex(std::string newVal)
+{
+    auto temp = std::find(mChoices.begin(), mChoices.end(), newVal);
+    if (temp == mChoices.end()) {
+        PPX_LOG_ERROR(mFlagName << " does not have this value in allowed range: " << newVal);
+        return;
+    }
+
+    // update mIndex and trigger mCallback
+    size_t newI = static_cast<size_t>(temp - mChoices.begin());
+    if (newI != mIndex) {
+        mIndex = newI;
+        if (mCallback)
+            mCallback(mIndex);
+    }
+}
+
+bool KnobStrDropdown::IsValidIndex(size_t i)
+{
+    if (static_cast<int>(i) < 0 || static_cast<int>(i) >= mChoices.size())
+        return false;
+    return true;
+}
+
+void KnobStrDropdown::SetDefault(size_t newI)
+{
+    PPX_ASSERT_MSG(IsValidIndex(newI), "invalid default index");
+    mDefaultIndex = newI;
+    ResetToDefault();
+}
+
+void KnobStrDropdown::SetDefault(std::string newVal)
+{
+    auto temp = std::find(mChoices.begin(), mChoices.end(), newVal);
+    PPX_ASSERT_MSG(temp != mChoices.end(), "invalid default value");
+
+    mDefaultIndex = static_cast<size_t>(temp - mChoices.begin());
+    ResetToDefault();
+}
+
+// -------------------------------------------------------------------------------------------------
+// Helper functions
+// -------------------------------------------------------------------------------------------------
+
+void DrawKnobs(const std::vector<Knob*>& knobPtrs)
```

Comment:
This could be removed if you drop the `all-knobs-have-children`, or refactorize the `draw` between a non-overridable `draw()`, and an overridable private `drawSelf()`:

```cpp
class Knob {
public:
    void Draw() {
        DrawSelf();
        for (auto& child : mChildren) { child.Draw(); };
    }

private:
    void DrawSelf() = 0;
}

class KnobSlider : final Knob {
private:
    void DrawSelf() {
        ImGui::DrawSlider(...);
    }
}
```

Reply:
There really should be a difference between a Node in the UI and a Knob